modules/qtwidgets/src/lightpositionwidgetqt.cpp

```diff
@@ -36,119 +42,100 @@
-    phi_ = atan2(y, x);
-    // Spherical to Cartesian coordinates
-    float x1 = sin(theta_) * cos(phi_) * gradientRadius;
-    float y1 = sin(theta_) * sin(phi_) * gradientRadius;
-    QPointF newPoint(x1, y1);
-    gradient_->setFocalPoint(newPoint + center);
-    gradientPixmap_->fill(Qt::transparent);
-    painter_->fillRect(0, 0, 100, 100, *gradient_);
-    this->setPixmap(*gradientPixmap_);
     emit positionChanged();
 }
 
-void LightPositionWidgetQt::setPosition(const vec3& p) {
+void LightPositionWidgetQt::setPosition(const dvec3& p) {
+    const auto rect = QRectF{pixmap_.rect()};
     radius_ = glm::length(p);
+    pos_ = dvec2{p.x, -p.y} / (radius_ == 0.0 ? 1.0 : radius_);
     radius_ *= p.z < 0.0f ? -1.0f : 1.0f;
 
-    if (radius_ == 0) return;
-    float gradientRadius = static_cast<float>(gradient_->radius());
-
-    QPointF center(gradientPixmap_->width() / 2, gradientPixmap_->height() / 2);
-    theta_ = acos(p.z / std::abs(radius_));
-    phi_ = atan2(-p.y, p.x);
-    float x1 = sin(theta_) * cos(phi_) * gradientRadius;
-    float y1 = sin(theta_) * sin(phi_) * gradientRadius;
-    float gradientSpaceRadius = sqrt(pow(x1, 2) + pow(y1, 2));
-
-    // Check if user clicked close to, or outside of radius
-    if (gradientSpaceRadius + 3.f > gradientRadius) {
-        float normFactor = gradientRadius / (3.f + gradientSpaceRadius);
-        x1 *= normFactor;
-        y1 *= normFactor;
-    }
-
-    QPointF newPoint(x1, y1);
-    gradient_->setFocalPoint(newPoint + center);
-    gradientPixmap_->fill(Qt::transparent);
-    painter_->fillRect(0, 0, 100, 100, *gradient_);
-    this->setPixmap(*gradientPixmap_);
+    const auto focal = rect.center() + 0.999 * gradient_.radius() * QPointF(pos_.x, pos_.y);
+    gradient_.setFocalPoint(focal);
+    pixmap_.fill(Qt::transparent);
+    QPainter painter(&pixmap_);
+    painter.fillRect(pixmap_.rect(), gradient_);
+    setPixmap(pixmap_);
 }
 
-vec3 LightPositionWidgetQt::getPosition() const {
-    return std::abs(radius_) * vec3(getX(), getY(), getZ());
+dvec3 LightPositionWidgetQt::getPosition() const {
+    return std::abs(radius_) * dvec3{getX(), getY(), getZ()};
 }
 
-void LightPositionWidgetQt::setRadius(float radius) {
+void LightPositionWidgetQt::setRadius(double radius) {
     radius_ = radius;
     emit positionChanged();
 }
 
-float LightPositionWidgetQt::getRadius() const { return radius_; }
+double LightPositionWidgetQt::getRadius() const { return radius_; }
 
-float LightPositionWidgetQt::getX() const { return sin(theta_) * cos(phi_); }
+double LightPositionWidgetQt::getX() const { return pos_.x; }
 
-float LightPositionWidgetQt::getY() const { return -sin(theta_) * sin(phi_); }
+double LightPositionWidgetQt::getY() const { return -pos_.y; }
 
-float LightPositionWidgetQt::getZ() const { return (radius_ < 0.0f ? -1.0f : 1.0f) * cos(theta_); }
+double LightPositionWidgetQt::getZ() const {
+    return (radius_ < 0.0f ? -1.0f : 1.0f) * std::sqrt(1.0 - glm::length(pos_));
```

Comment:
If you want to nitpick: remove f in -1.0f and 1.0f

Reply:
humm missed one... thanks